Mathlib/LinearAlgebra/QuadraticForm/Basic.lean

```diff
@@ -1030,66 +1084,78 @@ section Ring
+  fun x hx => smul_pos a_pos (h x hx)
+#align quadratic_form.pos_def.smul QuadraticMap.PosDef.smul
 
 variable {n : Type*}
 
-theorem PosDef.nonneg {Q : QuadraticForm R₂ M} (hQ : PosDef Q) (x : M) : 0 ≤ Q x :=
+theorem PosDef.nonneg {Q : QuadraticMap R₂ M N} (hQ : PosDef Q) (x : M) : 0 ≤ Q x :=
   (eq_or_ne x 0).elim (fun h => h.symm ▸ (map_zero Q).symm.le) fun h => (hQ _ h).le
-#align quadratic_form.pos_def.nonneg QuadraticForm.PosDef.nonneg
+#align quadratic_form.pos_def.nonneg QuadraticMap.PosDef.nonneg
 
-theorem PosDef.anisotropic {Q : QuadraticForm R₂ M} (hQ : Q.PosDef) : Q.Anisotropic := fun x hQx =>
-  by_contradiction fun hx =>
-    lt_irrefl (0 : R₂) <| by
+theorem PosDef.anisotropic {Q : QuadraticMap R₂ M N} (hQ : Q.PosDef) : Q.Anisotropic :=
+  fun x hQx => by_contradiction fun hx =>
+    lt_irrefl (0 : N) <| by
       have := hQ _ hx
       rw [hQx] at this
       exact this
-#align quadratic_form.pos_def.anisotropic QuadraticForm.PosDef.anisotropic
+#align quadratic_form.pos_def.anisotropic QuadraticMap.PosDef.anisotropic
 
-theorem posDef_of_nonneg {Q : QuadraticForm R₂ M} (h : ∀ x, 0 ≤ Q x) (h0 : Q.Anisotropic) :
+theorem posDef_of_nonneg {Q : QuadraticMap R₂ M N} (h : ∀ x, 0 ≤ Q x) (h0 : Q.Anisotropic) :
     PosDef Q := fun x hx => lt_of_le_of_ne (h x) (Ne.symm fun hQx => hx <| h0 _ hQx)
-#align quadratic_form.pos_def_of_nonneg QuadraticForm.posDef_of_nonneg
+#align quadratic_form.pos_def_of_nonneg QuadraticMap.posDef_of_nonneg
 
-theorem posDef_iff_nonneg {Q : QuadraticForm R₂ M} : PosDef Q ↔ (∀ x, 0 ≤ Q x) ∧ Q.Anisotropic :=
+theorem posDef_iff_nonneg {Q : QuadraticMap R₂ M N} : PosDef Q ↔ (∀ x, 0 ≤ Q x) ∧ Q.Anisotropic :=
   ⟨fun h => ⟨h.nonneg, h.anisotropic⟩, fun ⟨n, a⟩ => posDef_of_nonneg n a⟩
-#align quadratic_form.pos_def_iff_nonneg QuadraticForm.posDef_iff_nonneg
+#align quadratic_form.pos_def_iff_nonneg QuadraticMap.posDef_iff_nonneg
 
-theorem PosDef.add (Q Q' : QuadraticForm R₂ M) (hQ : PosDef Q) (hQ' : PosDef Q') :
+theorem PosDef.add (Q Q' : QuadraticMap R₂ M N) (hQ : PosDef Q) (hQ' : PosDef Q') :
     PosDef (Q + Q') := fun x hx => add_pos (hQ x hx) (hQ' x hx)
-#align quadratic_form.pos_def.add QuadraticForm.PosDef.add
-
-theorem linMulLinSelfPosDef {R} [LinearOrderedCommRing R] [Module R M] (f : M →ₗ[R] R)
-    (hf : LinearMap.ker f = ⊥) : PosDef (linMulLin f f) := fun _x hx =>
-  mul_self_pos.2 fun h => hx <| LinearMap.ker_eq_bot'.mp hf _ h
-#align quadratic_form.lin_mul_lin_self_pos_def QuadraticForm.linMulLinSelfPosDef
+#align quadratic_form.pos_def.add QuadraticMap.PosDef.add
+
+-- This should be in `Algebra/Ring/Defs` but that causes issues with `Tactic/Ring/Basic`
+-- see Note [lower instance priority]
+instance (priority := 100) _root_.CommSemiring.toNonUnitalNonAssocCommSemiring {N' : Type*}
+    [CommSemiring N'] : NonUnitalNonAssocCommSemiring N' :=
+  { inferInstanceAs (CommMonoid N'), inferInstanceAs (CommSemiring N') with }
```

Comment:
Can you split this to its own PR (in the right place), then ask for help with `ring` on zulip?

Reply:
Done #12617 and https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Adding.20an.20instance.20breaks.20.60Mathlib.2ETactic.2ERing.2EBasic.60

Reply:
We're restricting `linMulLin` to the `CommSemiring` case for now.